agent/lib/kontena/network_adapters/weave_executor.rb

```diff
@@ -1,155 +1,126 @@
-          retries += 1
-          error exc.message
-          sleep 0.5
-          retry if retries < 10
-
-          error exc.message
-          return false
-        end
+    def censor_password(command)
+      if command.include?('--password')
+        cmd = command.dup
+        passwd_index = cmd.index('--password')
+        cmd[passwd_index + 1] = '<redacted>'
 
-        status_code = response["StatusCode"]
-        output = container.streaming_logs(stdout: true, stderr: true)
-        command = censor_password(cmd)
-        if status_code != 0
-          error "weaveexec exit #{status_code}: #{command}\n#{output}"
-          return false
-        elsif block
-          debug "weaveexec stream: #{command}"
-          output.each_line &block
-          return true
-        else
-          debug "weaveexec ok: #{command}\n#{output}"
-          return true
-        end
-      ensure
-        container.delete(force: true, v: true) if container
+        cmd
+      else
+        command
       end
     end
 
-    # List network information for container(s)
-    #
-    # @param [Array<String>] what for given Docker IDs, 'weave:expose', or all
-    # @yield [name, mac, *cidrs]
-    # @yieldparam [Array<String>] cidrs
-    def ps(*what)
-      self.execute(['--local', 'ps', *what]) do |line|
-        yield *line.split()
+    # @param [Array<String>] *cmd
+    # @raise [Docker::Error]
+    # @raise [WeaveExecError]
+    # @yield [line] Each line of output
+    def run(*cmd, &block)
+      container = Docker::Container.create(
+        'Image' => IMAGE,
+        'Cmd' => ['--local'] + cmd,
+        'Volumes' => {
+          '/var/run/docker.sock' => {},
+          '/host' => {}
+        },
+        'Labels' => {
+          'io.kontena.container.skip_logs' => '1'
+        },
+        'Env' => [
+          'HOST_ROOT=/host',
+          "VERSION=#{WEAVE_VERSION}",
+          "WEAVE_DEBUG=#{WEAVE_DEBUG}",
+        ],
+        'HostConfig' => {
+          'Privileged' => true,
+          'NetworkMode' => 'host',
+          'PidMode' => 'host',
+          'Binds' => [
+            '/var/run/docker.sock:/var/run/docker.sock',
+            '/:/host'
+          ]
+        }
+      )
+      container.start!
+      response = container.wait
+
+      command = censor_password(cmd)
+      status_code = response["StatusCode"]
+
+      if status_code != 0
+        output = container.streaming_logs(stdout: true, stderr: true)
+        raise WeaveExecError.new(command, status_code, output)
+      elsif block
+        stderr = container.streaming_logs(stderr: true)
+        stdout = container.streaming_logs(stdout: true)
+        debug "weaveexec stream #{command}:\n#{stderr}"
+        stdout.each_line &block
+      else
+        output = container.streaming_logs(stdout: true, stderr: true)
+        debug "weaveexec ok: #{command}\n#{output}"
       end
+    ensure
+      # XXX: rescue cleanup errors
+      container.delete(force: true, v: true) if container
```

Comment:
> I think we should get rid of that pool.

Howcome? I think the celluloid pool is still useful to limit how many weaveexecs get run in parallel, and also prevent the weave actors from blocking... I think it might be important for the `Kontena::Launchers::Weave` to remain responsive to `Observable` requests while it's busy launching weave?